src/parquet/arrow/reader.cc

```diff
@@ -1159,5 +1328,102 @@ Status ColumnReader::NextBatch(int batch_size, std::shared_ptr<Array>* out) {
   return impl_->NextBatch(batch_size, out);
 }
 
+// StructImpl methods
+
+Status StructImpl::CreateBitmap(size_t num_elements,
+    std::shared_ptr<PoolBuffer>* null_bitmap_out) {
+  *null_bitmap_out = std::make_shared<::arrow::PoolBuffer>(pool_);
+  auto bitmap_size = ::arrow::BitUtil::CeilByte(num_elements) / 8;
+  RETURN_NOT_OK((*null_bitmap_out)->Resize(bitmap_size));
+  uint8_t* null_bitmap_ptr = (*null_bitmap_out)->mutable_data();
+  memset(null_bitmap_ptr, 0xFF, bitmap_size);
+  return Status::OK();
+}
+
+Status StructImpl::DefLevelsToNullArray(
+    std::shared_ptr<PoolBuffer>* null_bitmap_out,
+    int64_t* null_count_out) {
+  std::shared_ptr<PoolBuffer> null_bitmap;
+  auto null_count = 0;
+  ValueLevelsPtr def_levels_data;
+  size_t def_levels_length;
+  RETURN_NOT_OK(GetDefLevels(&def_levels_data, &def_levels_length));
+  RETURN_NOT_OK(CreateBitmap(def_levels_length, &null_bitmap));
+  uint8_t* null_bitmap_ptr = null_bitmap->mutable_data();
+  for (size_t i = 0; i < def_levels_length; i++) {
+    if (def_levels_data[i] < struct_def_level_) {
+      // Mark null
+      null_count += 1;
+      ::arrow::BitUtil::ClearBit(null_bitmap_ptr, i);
+    }
+  }
+
+  *null_count_out = null_count;
+  *null_bitmap_out = (null_count == 0) ? nullptr : null_bitmap;
+  return Status::OK();
+}
+
+// TODO(itaiin): Consider caching the results of this calculation -
+//   note that this is only used once for each read for now
+Status StructImpl::GetDefLevels(ValueLevelsPtr* data, size_t* length) {
+  *data = nullptr;
+  if (children_.size() == 0) {
+    // Empty struct
+    *length = 0;
+    return Status::OK();
+  }
+
+  // We have at least one child
+  ValueLevelsPtr child_def_levels;
+  size_t child_length;
+  RETURN_NOT_OK(children_[0]->GetDefLevels(&child_def_levels, &child_length));
+  auto size = child_length * sizeof(int16_t);
+  def_levels_buffer_.Resize(size);
+  // Initialize with the minimal def level
+  std::memset(def_levels_buffer_.mutable_data(), 0, size);
+  auto result_levels = reinterpret_cast<int16_t*>(def_levels_buffer_.mutable_data());
+
+  // A struct is defined if at least one of its children is defined,
+  // so the definition level of a struct value is the max definition level
+  // of its children
```

Comment:
Its true. The `max` logic is still correct (given the def data isn't botched somehow) but I have to make the first line of the comment more accurate since it neglects that scenario.